### PR TITLE
api: get my open/closed orders on all markets

### PR DESCRIPTION
**Pull request by lightsing**

### CODE REVIEW

File: src/matchengine/controller.rs [hunk 1]
```diff
@@ -219,24 +220,28 @@ impl Controller {
...
+        let orders_by_market: Vec<Box<dyn Iterator<Item = Order>>> = markets
+            .map(|m| {
+                m.users
+                    .get(&req.user_id)
+                    .map(|order_map| Box::new(order_map.values().rev().map(|order_rc| order_rc.deep())) as Box<dyn Iterator<Item = Order>>)
+                    .unwrap_or_else(|| Box::new(Vec::new().into_iter()) as Box<dyn Iterator<Item = Order>>)
             })
-            .unwrap_or_else(Vec::new);
+            .collect();
+        let orders = MergeSortIterator::compare_by(orders_by_market, SortOrder::Asc, |a, b| a.id.cmp(&b.id))
```
Comment (HAOYUatHZ) [inline src/matchengine/controller.rs:240]: compare by `create_time`?

File: examples/js/multi_market.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        ));
+    console.log(orders);
+
+    const openOrders = (await axios.get(`http://${server}/api/orders/all/1`)).data;
+    console.log(openOrders);
+
+    await Promise.all(orders.map(([market, id]) => client.orderCancel(1, market, Number(id))));
+
+    const closedOrders = (await axios.get(`http://${server}/restapi/closedorders/all/1`)).data;
+    console.log(closedOrders);
```
Comment (HAOYUatHZ) [inline examples/js/multi_market.ts:52]: we need to assert instead of only logging

**Review by HAOYUatHZ — CHANGES_REQUESTED** [review]
see comments above

File: src/matchengine/controller.rs (hunk 1 above)
Comment (lightsing) [reply src/matchengine/controller.rs:240]: it was sort by id in original version.

File: src/matchengine/controller.rs [hunk 3]
```diff
@@ -219,24 +220,28 @@ impl Controller {
...
-                    .rev()
-                    .skip(req.offset as usize)
-                    .take(limit as usize)
-                    .map(|order_rc| OrderInfo::from(order_rc.deep()))
-                    .collect()
+            .iter()
+            .filter(|(key, _market)| req.market == "all" || req.market == **key)
+            .map(|(_key, market)| market);
+        let total_order_count: usize = markets
+            .clone()
```
Comment (lispc) [inline src/matchengine/controller.rs:229]: Will this clone trigger a lot of memory copy？

Comment (lispc) [reply src/matchengine/controller.rs:229]: Full order book clone ？

File: src/matchengine/controller.rs (hunk 1 above)
Comment (HAOYUatHZ) [reply src/matchengine/controller.rs:240]: but in https://github.com/Fluidex/dingir-exchange/issues/231 @lispc says sort by `create_time`?

Comment (lightsing) [reply src/matchengine/controller.rs]: revert commit  79f675d

File: src/matchengine/controller.rs (hunk 3 above)
Comment (lightsing) [reply src/matchengine/controller.rs:229]: Clone the iterator will not clone the HashMap

**Review by HAOYUatHZ — DISMISSED** [review]

**Review by HAOYUatHZ — APPROVED** [review]

**Review by HAOYUatHZ — APPROVED** [review]

**Pull request merged**